gapic/schema/wrappers.py

```diff
@@ -1067,6 +1067,66 @@ def try_parse_routing_rule(cls, routing_rule: routing_pb2.RoutingRule) -> Option
         params = [RoutingParameter(x.field, x.path_template) for x in params]
         return cls(params)
 
+    @classmethod
+    def resolve(cls, routing_rule: routing_pb2.RoutingRule, request: Union[dict, str]) -> dict:
+        """Resolves the routing header which should be sent along with the request.
+        This function performs dynamic header resolution, identical to what's in `client.py.j2`.
```

Comment:
Where, exactly, is it identical? It seems like if it is identical it should move to a shared library, like api_core.

Also, I'm not clear about how `wrappers.py` gets propagated to the templates. Does this comment mean it isn't?

Reply:
> Where, exactly, is it identical? 

It is actually in `_client_macros.j2` not `client.py.j2`. The code was copied from the test below

https://github.com/googleapis/gapic-generator-python/blob/4c5de8791795f8101f6ec66f80b8a8e5e9a21822/tests/unit/schema/wrappers/test_routing.py#L27

I've updated the comment and included a link to the specific code in `_client_macros.j2` which is here

https://github.com/googleapis/gapic-generator-python/blob/4c5de8791795f8101f6ec66f80b8a8e5e9a21822/gapic/templates/%25namespace/%25name_%25version/%25sub/services/%25service/_client_macros.j2#L150-L164


> It seems like if it is identical it should move to a shared library, like api_core.
Good point. I've filed https://github.com/googleapis/gapic-generator-python/issues/2160. I didn't make the change in this PR as this is a refactor of `_client_macros.j2`  which is not the purpose of this PR. This PR is improving the assert and moving logic which already exists in `tests/unit/schema/wrappers/test_routing.py` to `gapic/schema/wrappers.py` so that the same code that we use in gapic-generator tests can also be used in generated libraries.

> Also, I'm not clear about how wrappers.py gets propagated to the templates. Does this comment mean it isn't?

We can call the helper functions in wrappers.py from templates but we cannot use it to generate client code. For example, see the code below where we call the helper function `method.routing_rule.resolve`.

https://github.com/googleapis/gapic-generator-python/blob/e50bbd6921ba3dc112c5c625c5f16fc7d7f6f57c/gapic/templates/tests/unit/gapic/%25name_%25version/%25sub/test_macros.j2#L428

https://github.com/googleapis/gapic-generator-python/issues/2160 is for refactoring  `_client_macros.j2` to use helper functions to build the explicit routing headers in `google-api-core` rather than having the explicit routing header logic in templates.

Reply:
Ah, thanks!